### PR TITLE
fix: promote `resid_prototype` using tunables

### PR DESCRIPTION
**Pull request by AayushSabharwal**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

**Review by ChrisRackauckas — APPROVED** [review]

File: src/systems/nonlinear/nonlinearsystem.jl [hunk 1]
```diff
@@ -295,7 +295,7 @@ function SciMLBase.NonlinearFunction(sys::NonlinearSystem, args...; kwargs...)
 end
 
 function SciMLBase.NonlinearFunction{iip}(sys::NonlinearSystem, dvs = unknowns(sys),
-        ps = parameters(sys), u0 = nothing;
+        ps = parameters(sys), u0 = nothing, p = nothing;
```
Comment (ChrisRackauckas) [inline src/systems/nonlinear/nonlinearsystem.jl:298]: Why ps and p here?

Comment (AayushSabharwal) [reply src/systems/nonlinear/nonlinearsystem.jl:298]: ps is the list of parameter symbols. p is the parameter object, which is required for type promotion in resid_prototype.

Comment (ChrisRackauckas) [reply src/systems/nonlinear/nonlinearsystem.jl:298]: How is p already generated by here?

Comment (AayushSabharwal) [reply src/systems/nonlinear/nonlinearsystem.jl:298]: It's generated in `process_NonlinearProblem`, which calls this and is responsible for providing `u0` and now `p` as well.

**Pull request merged**